### PR TITLE
[WIP] Add PSI memory statistics collector and visualization tool.

### PR DESCRIPTION
**Pull request by OhmSpectator**
This patch set introduces a memory pressure statistics collector and a visualization.

**Note**: The memory pressure statistics collector and the associated tools are not intended to be integrated into the EVE system as a permanent feature. Instead, these tools are designed to help understand memory pressure behavior, which will be used to set appropriate thresholds in the monitor tool later.

The new functionality includes the following key components:

**Memory Pressure Statistics Collector** is a daemon implemented to collect memory pressure statistics from `/proc/pressure/memory` every second.
The collected data is written to an output file (`psi_stats.txt`) in the `/persist/memory-monitor/output` directory. The output file includes detailed headers and columns for all relevant data fields. The collector ensures data is written immediately using fsync to prevent data loss in the event of a system crash. The process is daemonized to run in the background, ensuring continuous monitoring.

**Visualization Tool (`visualize.py`)** is provided to visualize the collected PSI statistics.
The visualization script reads the log file, combines the date and time columns into a single timestamp, and generates interactive plots using Plotly. The plots display both the average and total memory pressure metrics over time, with the totals shown as cumulative area plots on a secondary y-axis. A new Makefile target (`local-view-psi-stats`) is included to facilitate running the visualization tool. This target also handles installing the required Python packages (pandas and plotly) specified in requirements.txt.

Also, Makefile and README are updated to reflect the latest changes and provide a detailed explanation.

### CODE REVIEW

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    if (fsync(fd) == -1) {
+        perror("fsync");
+        exit(EXIT_FAILURE);
+    }
+}
+
+static void daemonize(void)
+{
+    int pid, sid;
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:28]:
```suggestion
    int sid;
    pid_t pid;
```

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    // If we got a good PID, then we can exit the parent process
+    if (pid > 0) {
+        exit(EXIT_SUCCESS);
+    }
+
+    // Change the file mode mask
+    umask(0);
+
+    // Create a new SID for the child process
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:41]: Why is this necessary?

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    pid = fork();
+    if (pid < 0) {
+        exit(EXIT_FAILURE);
+    }
+    // If we got a good PID, then we can exit the parent process
+    if (pid > 0) {
+        exit(EXIT_SUCCESS);
+    }
+
+    // Change the file mode mask
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:38]: Why is this necessary?

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    // If /persist directory exists, create the output file there, otherwise create it in the current directory
+    // But first, create the filename dynamically, in a static size buffer
+    char output_file[PATH_MAX];
+    if (access(DEVICE_OUTPUT_DIR, F_OK) == 0) {
+        snprintf(output_file, PATH_MAX, "%s/%s", DEVICE_OUTPUT_DIR, OUTPUT_FILE);
+    } else {
+        snprintf(output_file, PATH_MAX, "./%s", OUTPUT_FILE);
+    }
+
+    fd = open(output_file, O_WRONLY | O_CREAT | O_TRUNC, 0644);
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:68]:
```suggestion
    fd = open(output_file, O_WRONLY | O_CREAT | O_TRUNC, 0644);
    if (fd < 0) ...
```

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <time.h>
+#include <unistd.h>
+
+#define BUFFER_SIZE 256
+#define OUTPUT_FILE "psi_stats.txt"
+#define DEVICE_OUTPUT_DIR "/persist/memory-monitor/output"
+#define PROC_FILE "/proc/pressure/memory"
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:13]:
```suggestion
#define MEMORY_PRESSURE_FILE "/proc/pressure/memory"
```

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 2 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:41]: To be detached from the terminal. So we can strat the process in an SSH session and close the terminal, leaving the process collecting the statistics.

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            perror("fopen");
+            close(fd);
+            exit(EXIT_FAILURE);
+        }
+
+        char some_avg10[16], some_avg60[16], some_avg300[16], some_total[16];
+        char full_avg10[16], full_avg60[16], full_avg300[16], full_total[16];
+
+        while (fgets(buffer, BUFFER_SIZE, proc_file) != NULL) {
+            sscanf(buffer, "some avg10=%s avg60=%s avg300=%s total=%s",
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:87]:
```suggestion
            sscanf(buffer, "some avg10=%16s avg60=%16s avg300=%16s total=%16s",
```
or should it be 15, because it adds the `\0`?

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                   some_avg10, some_avg60, some_avg300, some_total);
+            fgets(buffer, BUFFER_SIZE, proc_file);
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:89]: Check the return value?

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 1 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:28]: Do we have some code standard for that kind of things? If so, please share it so I can check before making PRs.

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 8]
```diff
@@ -0,0 +1,108 @@
+#include <fcntl.h>
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:1]: Copyright?

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 3 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:38]: This way, I can create the log files with proper permissions regardless of how I start the process (who is the parent). Maybe it's not necessary now, but it does not hurt. Be independent from the parent!

File: pkg/memory-monitor/Makefile [hunk 9]
```diff
@@ -32,7 +32,15 @@ $(BIN_DIR)/pressure: $(PRESSURE_OBJ)
...
 pressure: $(BIN_DIR)/pressure
 
-TARGETS=memory-monitor pressure
+STATISTICS_COLLECTOR_DIR=$(SRC_DIR)/statistics-collector
+STATISTICS_COLLECTOR_SRC=$(wildcard $(STATISTICS_COLLECTOR_DIR)/*.c)
+STATISTICS_COLLECTOR_OBJ=$(patsubst $(SRC_DIR)/%.c,$(OBJ_DIR)/%.o,$(STATISTICS_COLLECTOR_SRC))
+$(BIN_DIR)/statistics-collector: $(STATISTICS_COLLECTOR_OBJ)
+	@mkdir -p $(BIN_DIR)
+	$(CC) $(CFLAGS) -o $@ $^
+statistics-collector: $(BIN_DIR)/statistics-collector
```
Comment (christoph-zededa) [inline pkg/memory-monitor/Makefile:41]: How will it be cross-compiled?
1. from amd64 to arm64
2. from glibc to musl
3. from OS X to Linux

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 4 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:68]: Yep, will fix

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 5 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:13]: Okaaay

File: pkg/memory-monitor/README.md [hunk 10]
```diff
@@ -354,3 +354,44 @@ It will allocate 100Mb of memory and release it after user presses Enter.
...
+pressure.
+
+The tool is built using the Makefile:
+
+```shell
+$ make statistics-collector
+```
+
+It will create a `statistics-collector` binary in the `bin` directory. The
+binary can be copied to the EVE system and run from the /persist directory.
```
Comment (christoph-zededa) [inline pkg/memory-monitor/README.md:372]: Would be good to add a description on how to cross-compile it.

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 4 above)
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:68]:
```suggestion
    fd = open(output_file, O_WRONLY | O_CREAT | O_TRUNC | O_DIRECT | O_SYNC, 0644);
```
so that you don't have to `fsync` all the time?

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 6 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:87]: Good suggestion. For avgN the max value is "100.00". So the buffer will contain 7 bytes (and I have to specify 6 in the format string). I can fix the sizes. For the totals, I would set the lens to 20 (and 19 for format) to fit max long long.

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    if (sid < 0) {
+        exit(EXIT_FAILURE);
+    }
+}
+
+int main()
+{
+    int fd;
+    FILE *proc_file;
+    char buffer[BUFFER_SIZE];
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:52]:
```suggestion
    char pressure_file_buffer[BUFFER_SIZE];
```

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 8 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:1]: Oops, will add. I though I had already done it.

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            fgets(buffer, BUFFER_SIZE, proc_file);
+            sscanf(buffer, "full avg10=%s avg60=%s avg300=%s total=%s",
+                   full_avg10, full_avg60, full_avg300, full_total);
+
+            time(&rawtime);
+            timeinfo = localtime(&rawtime);
+            char timestamp[BUFFER_SIZE];
+            strftime(timestamp, BUFFER_SIZE, "%Y-%m-%d %H:%M:%S", timeinfo);
+
+            dprintf(fd, "%s %s %s %s %s %s %s %s %s\n", timestamp, some_avg10, some_avg60, some_avg300, some_total, full_avg10, full_avg60, full_avg300, full_total);
```
Comment (christoph-zededa) [inline pkg/memory-monitor/src/statistics-collector/main.c:98]:
```suggestion
            dprintf(fd, "%s, %s, %s, %s, %s, %s, %s, %s, %s\n", timestamp, some_avg10, some_avg60, some_avg300, some_total, full_avg10, full_avg60, full_avg300, full_total);
```
then it is a "real" CSV and you don't have to set the separator explicitly?

File: pkg/memory-monitor/README.md (hunk 10 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/README.md:372]: To be honest, I didn't think about cross-compiling. But it's a problem for the other binaries here as well. Good point. I would prefer to create a task and address it later.

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 4 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:68]: True. Can use this approach.

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 12 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:98]: Do I have to set it now?

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 2 above)
Comment (christoph-zededa) [reply pkg/memory-monitor/src/statistics-collector/main.c:41]: I thought `setsid` just makes the process to the leader of the process group.
I don't think it installs a `SIGHUP` handler and closes stdin/stdout/stderr.

According to the `man 2 setsid` you still get SIGHUP:
> If  a session has a controlling terminal, and the CLOCAL flag for that terminal is not set, and a terminal hangup > occurs, then the session leader is sent a SIGHUP signal.

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 11 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:52]: Okaaay

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 1 above)
Comment (christoph-zededa) [reply pkg/memory-monitor/src/statistics-collector/main.c:28]: `pid_t` comes from `man 2 fork`.

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 2 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:41]: True, but in the same man 2:
```
Initially, the new session has no controlling terminal.
```

File: pkg/memory-monitor/src/statistics-collector/main.c (hunk 1 above)
Comment (OhmSpectator) [reply pkg/memory-monitor/src/statistics-collector/main.c:28]: Aaaah, I see, sorry

**Comment by OhmSpectator** [comment]
In general, after a discussion with @christoph-zededa, it looks like the same functionality, but already integrated in EVE and available as a command, can be easily added as a Go function in `agentlog.go`. I'll think about it.

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+#include <time.h>
+#include <unistd.h>
+
+#define BUFFER_SIZE 256
+#define OUTPUT_FILE "psi_stats.txt"
+#define DEVICE_OUTPUT_DIR "/persist/memory-monitor/output"
+#define PROC_FILE "/proc/pressure/memory"
+
+void write_to_file(int fd, const char *buffer) {
+    if (write(fd, buffer, strlen(buffer)) == -1) {
```
Comment (rene) [inline pkg/memory-monitor/src/statistics-collector/main.c:16]: write() can return different errors, so better to check for `write(...) < 0`

File: pkg/memory-monitor/src/statistics-collector/main.c [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        perror("write");
+        exit(EXIT_FAILURE);
+    }
+    if (fsync(fd) == -1) {
```
Comment (rene) [inline pkg/memory-monitor/src/statistics-collector/main.c:20]: same for fsync()

**Comment by OhmSpectator** [comment]
I've reimplemented the tool in Go so that it's available in Pillar and as a standalone tool. I will create a new PR later.
